### PR TITLE
CRM457-1281 - build test data

### PR DESCRIPTION
**Pull request by dwhenry**
## Description of change
Build NSM data for performance testing

 [Link to relevant ticket](https://dsdmoj.atlassian.net/browse/CRM457-1281)

### CODE REVIEW

File: lib/test_data/nsm_builder.rb [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        args, kwargs = *options(**options).values.sample
+        claim = FactoryBot.create(*args, kwargs.call)
+
+        invalid_tasks = check_tasks(claim)
+
+        if invalid_tasks.any?
+          Rails.logger.debug { "Invalid for #{invalid_tasks.map(&:last).join(', ')}" }
+
+          # rubocop:disable Lint/Debugger
+          debugger if ENV['DEBUG_TEST_DATA']
```
Comment (dwhenry) [inline lib/test_data/nsm_builder.rb:22]: should I remove this as was super useful when debugging why create failed?

File: spec/factories/disbursements.rb [hunk 2]
```diff
@@ -8,7 +8,7 @@
     trait :valid do
       disbursement_date { age.days.ago.to_date }
       disbursement_type { DisbursementTypes.values.reject(&:other?).sample.to_s }
-      miles { 50 + (200**Random.new.rand(1.0).to_i) } # range: 50 -> 250
+      miles { 50 + (200**Random.new.rand(1.0).to_i) } # range: 50 -> 250 with logorithmic distribution
```
Comment (patrick-laa) [inline spec/factories/disbursements.rb:11]: I'm not entirely sure what you mean by logarithmic distribution here, but I don't think this code does what you think it does. `Random.new.rand(1.0)` is always less than 1, so `Random.new.rand(1.0).to_i` is always 0, so `(200**Random.new.rand(1.0).to_i)` is always 1, so this always sets miles to 51

Comment (patrick-laa) [reply spec/factories/disbursements.rb:11]: Ah, ok, I see what you mean by logarithmic distribution. You just need to move the `)` that's currently to the right of `to_i` the left of it.

Comment (dwhenry) [reply spec/factories/disbursements.rb:11]: I don't understand why the wrote the rand function like this... I realised halfway through that the 1.0 broke things and thought I had removed it from everywhere... :(

fixed now

also for reference:
```
irb(main):001> Random.new.rand(1.0)
=> 0.9108281877297423
irb(main):002> rand(1.0)
=> 0
```

File: lib/test_data/nsm_builder.rb [hunk 3]
```diff
@@ -0,0 +1,86 @@
+module TestData
+  class NsmBuilder
+    def build_many(bulk: 100, large: 4, year: 2023)
+      raise 'Do not run on production' if HostEnv.production?
+
+      bulk.times do
+        build(year:)
+
+        # avoid issues with large number of applications with the same last_updated_at time
+        sleep 0.1 unless HostEnv.local?
```
Comment (patrick-laa) [inline lib/test_data/nsm_builder.rb:10]: Why is this not an issue locally?

File: lib/test_data/nsm_builder.rb [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+      end
+
+      Rails.logger.info "Created large examples: #{large_ids.to_sentence}"
+    end
+
+    def build(**options)
+      ActiveRecord::Base.transaction do
+        args, kwargs = *options(**options).values.sample
+        claim = FactoryBot.create(*args, kwargs.call)
+        # Nsm::Steps::SolicitorDeclarationForm.new(application: claim, record: claim, signatory_name: Faker::Name.name).save
```
Comment (patrick-laa) [inline lib/test_data/nsm_builder.rb:27]: Guessing this can be deleted?

File: lib/test_data/nsm_builder.rb [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+          end
+        ],
+        breach: [
+          [:claim, :complete, :case_type_breach, :build_associates],
+          proc do
+            date = date_for(year)
+            { date: date, disbursements_count: rand(max / 2), work_items_count: rand(min..max), updated_at: date }
+          end
+        ],
+        no_disburesments: [
```
Comment (patrick-laa) [inline lib/test_data/nsm_builder.rb:64]: typo `disbursements`

**Review by patrick-laa — APPROVED** [review]

File: lib/test_data/nsm_builder.rb (hunk 3 above)
Comment (dwhenry) [reply lib/test_data/nsm_builder.rb:10]: so this is mostly to avoid the delay in the tests. I will change it to `Rails.env.test?` so this is more clear

**Pull request merged**